File: overlay/tests/test_cas.py
"""
test_cas.py - Test compare-and-swap (CAS) protocol operations.
"""

import os
import sys
import unittest

sys.path.insert(0, os.path.dirname(__file__))
from memcached_harness import MemcachedServer


class TestCas(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.server = MemcachedServer().start()

    @classmethod
    def tearDownClass(cls):
        cls.server.stop()

    def setUp(self):
        self.mc = self.server.client()

    def tearDown(self):
        self.mc.close()

    def test_cas_workflow(self):
        # Set initial item
        self.assertEqual(self.mc.set("cas_key", "val1"), "STORED")

        # Fetch with gets to obtain CAS id
        res = self.mc.gets("cas_key")
        self.assertIn("cas_key", res)
        val, flags, cas_id = res["cas_key"]
        self.assertEqual(val, "val1")
        self.assertTrue(cas_id > 0)

        # Successful CAS
        self.assertEqual(self.mc.cas("cas_key", "val2", cas_id), "STORED")
        val, flags, cas_id2 = self.mc.gets("cas_key")["cas_key"]
        self.assertEqual(val, "val2")
        self.assertNotEqual(cas_id, cas_id2)

        # Stale CAS should fail with EXISTS
        self.assertEqual(self.mc.cas("cas_key", "val3", cas_id), "EXISTS")

        # Normal set overwrites and updates CAS
        self.assertEqual(self.mc.set("cas_key", "val4"), "STORED")
        val, flags, cas_id3 = self.mc.gets("cas_key")["cas_key"]
        self.assertEqual(val, "val4")

        # CAS on non-existing key returns NOT_FOUND
        self.assertEqual(self.mc.cas("nonexistent_cas", "v", 12345), "NOT_FOUND")


if __name__ == "__main__":
    unittest.main()

File: tests/test_cas.c
/*
 * test_cas.c - Test compare-and-swap (CAS) protocol operations.
 */

#include "c_harness.h"

static int test_cas_workflow(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char resp[64];
  char val[64];
  test_uint64_t cas1 = 0;
  test_uint64_t cas2 = 0;

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_set(&c, "cas_key", "val1", 4, 0, 0, resp, sizeof(resp)) ==
              0);
  TEST_ASSERT_STR_EQ(resp, "STORED");

  TEST_ASSERT(client_gets(&c, "cas_key", val, sizeof(val), NULL, &cas1) == 0);
  TEST_ASSERT_STR_EQ(val, "val1");
  TEST_ASSERT(cas1 > 0);

  TEST_ASSERT(client_cas(&c, "cas_key", "val2", 4, cas1, 0, 0, resp,
                         sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "STORED");

  TEST_ASSERT(client_gets(&c, "cas_key", val, sizeof(val), NULL, &cas2) == 0);
  TEST_ASSERT_STR_EQ(val, "val2");
  TEST_ASSERT(cas2 > 0);
  TEST_ASSERT(cas2 != cas1);

  TEST_ASSERT(client_cas(&c, "cas_key", "val3", 4, cas1, 0, 0, resp,
                         sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "EXISTS");

  TEST_ASSERT(client_get(&c, "cas_key", val, sizeof(val), NULL) == 0);
  TEST_ASSERT_STR_EQ(val, "val2");

  TEST_ASSERT(client_set(&c, "cas_key", "val4", 4, 0, 0, resp, sizeof(resp)) ==
              0);
  TEST_ASSERT_STR_EQ(resp, "STORED");
  TEST_ASSERT(client_get(&c, "cas_key", val, sizeof(val), NULL) == 0);
  TEST_ASSERT_STR_EQ(val, "val4");

  TEST_ASSERT(client_cas(&c, "nonexistent_cas", "v", 1, 12345, 0, 0, resp,
                         sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "NOT_FOUND");

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

int test_cas_run(void) {
  RUN_TEST(test_cas_workflow);
  return 0;
}
